lib/spack/spack/version.py

```diff
@@ -886,3 +980,170 @@ class VersionError(spack.error.SpackError):
+            return self.data
+
+        # Were we given an already cloned path?
+        already_cloned = os.path.exists(self.repository_uri)
+        if already_cloned:
+            dest = self.repository_uri
+
+        # Otherwise honor the namespace for the service and repository
+        else:
+            dest = os.path.join(spack.paths.user_repos_cache_path, self.repository_uri)
+
+        # prepare a cache for the repository and metadata
+        for path in [dest, self.repository_metadata_path]:
+            parent = os.path.dirname(path)
+            if not os.path.exists(parent):
+                os.makedirs(parent)
+
+        # Only clone if we don't have it!
+        if not os.path.exists(dest):
+            self.fetcher.clone(os.path.dirname(dest))
+
+        # But if we get here and no repostiory, this is a fail
+        if not os.path.exists(dest):
+            tty.die("There was an issue cloning %s to %s" % (self.fetcher.url, dest))
+
+        # Always pull and fetch
+        with working_dir(dest):
+
+            # Don't do anything if already cloned - it might not have remote
+            if not already_cloned:
+                self.fetcher.git("pull")
+                self.fetcher.git("fetch")
+
+            # List tags (refs) by date, so last reference of a tag is newest
+            tags = self.fetcher.git("for-each-ref", "--sort=creatordate", "--format",
+                                    "%(refname)", "refs/tags", output=str).split('\n')
+
+        # Lookup of spack tags to git references
+        tag_lookup = {}
+
+        # Lookup of commits to spack versions
+        commit_to_version = {}
+
+        # For each tag, try to match to a version.
+        versions = [str(v) for v in spack_versions]
+        for tag in tags:
+            tag = tag.replace('refs/tags/', '', 1)
+            for v in versions:
+                if v in tag:
+                    tag_lookup[v] = tag
+
+        # Also prepare to parse versions that aren't in spack
+        for tag in tags:
+            tag = tag.replace('refs/tags/', '', 1)
+
+            # Try to derive a semantic version
+            match = SEMVER_REGEX.match(tag)
+            if match:
+                semver = match.groupdict()['semver']
+                if semver not in tag_lookup:
+                    tag_lookup[semver] = tag
+
+        # Now get commits for these versions
+        with working_dir(dest):
+            for spack_version, tag in tag_lookup.items():
+                command = ["rev-list", "-n", "1", "tags/%s" % tag]
+                commit = self.fetcher.git(*command, output=str)
+                commit_to_version[commit.strip()] = spack_version
+
+            # Get list of all commits, this is in reverse order
+            commits = self.fetcher.git("log", "--all", "--pretty=format:%H", output=str)
+
+        # commits[0] is most recent, commits[-2] is first
+        commits = [c for c in commits.split('\n') if c]
+
+        for commit in commits:
```

Comment:
This looks to me like it works, but it looks like it'll be very slow for large repositories -- I feel like there should be a way to do this that makes O(tags) calls to `git`, but it is not coming to me off the top of my head.

If this works then I say we get it in and optimize later -- would like to see more tests with more branch structure to ensure that (see comments above on the mock repo fixture)